aten/src/ATen/Exceptions.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+#include <typeinfo>
+
+
+namespace at {
+namespace checks {
+
+inline bool tensor_on_same_device(const at::Tensor &ten1, const at::Tensor &ten2){
+  return ten1.get_device()==ten2.get_device();
+}
+
+inline bool tensor_on_same_device(
+  const at::Tensor &ten1,
+  const c10::optional<at::Tensor> &ten2
+){
+  return !ten2.has_value() || ten1.get_device()==ten2->get_device();
+}
+
+inline bool tensor_on_cpu(const at::Tensor &ten){
+  return !ten.is_cuda(); // TODO: Should be a better way to do this
+}
+
+inline bool tensor_on_cpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || !ten->is_cuda(); // TODO: Should be a better way to do this
+}
+
+inline bool tensor_on_cuda_gpu(const at::Tensor &ten){
+  return ten.is_cuda();
+}
+
+inline bool tensor_on_cuda_gpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || ten->is_cuda();
```

Comment:
Interesting. So if the optional is empty this will return true?

Reply:
That's correct. The context in which we've been using these begins each function with a series of precondition checks including a `TENSOR_ON_GPU(optional_tensor)`. It might be better to break this out into a `TENSOR_ON_GPU()` for actual tensors and `OPTIONAL_TENSOR_NULL_OR_ON_GPU()` for the optional case, but this seemed potentially unnecessarily verbose.